r3bbase/R3BTprevTnext.cxx

```diff
@@ -0,0 +1,168 @@
+        return kFATAL;
+    }
+
+    fSamplerMapped = dynamic_cast<TClonesArray*>(rootManager->GetObject("SamplerMapped"));
+    if (fSamplerMapped == nullptr)
+    {
+        LOG(error) << "SamplerMapped not found or wrong type.";
+        return kFATAL;
+    }
+
+    fSamplerMSMapped = dynamic_cast<TClonesArray*>(rootManager->GetObject("SamplerMSMapped"));
+    if (fSamplerMSMapped == nullptr)
+    {
+        LOG(error) << "SamplerMSMapped not found or wrong type.";
+        return kFATAL;
+    }
+
+    fMSOffsetPar = dynamic_cast<R3BMSOffsetPar*>(rtdb->getContainer("MSOffsetPar"));
+    if (fMSOffsetPar == nullptr)
+    {
+        LOG(error) << "Could not find MSOffsetPar container!";
+        return kFATAL;
+    }
+
+    return kSUCCESS;
+}
+
+void R3BTprevTnext::Exec(Option_t* /*opt*/)
+{
+    Double_t fTprev = INVALID_TPTN; // Initialization with invalid values.
+    Double_t fTnext = INVALID_TPTN;
+    Int_t dMScounter = 0;
+    const Int_t sampHits = fSamplerMapped->GetEntriesFast();
+    const Int_t sampmsHits = fSamplerMSMapped->GetEntriesFast();
+    const Double_t MSOffset = fMSOffsetPar->GetMSOffset();
+    Double_t SAMPTime = 0;
+    Double_t SAMPMSTime = 0;
+    if (sampmsHits == 1 && sampHits > 0)
+    {
+        R3BSamplerMappedData* SAMPMapped = nullptr;
+        R3BSamplerMappedData* SAMPMSMapped = nullptr;
+        SAMPMSMapped = dynamic_cast<R3BSamplerMappedData*>(fSamplerMSMapped->At(0));
+        SAMPMSTime = SAMPMSMapped->GetTime();
+        for (Int_t i = 0; i < sampHits; ++i)
+        {
+            SAMPMapped = dynamic_cast<R3BSamplerMappedData*>(fSamplerMapped->At(i));
+            SAMPTime = SAMPMapped->GetTime();
+            auto tpn = SAMPTime - SAMPMSTime - MSOffset;
+            if (tpn < -fDelta_clk)
+            {
+                fTprev = -CLOCK_PERIOD * tpn;
+            }
+            if (abs(tpn) <= fDelta_clk)
+            {
+                ++dMScounter;
+            }
+            if (tpn > fDelta_clk)
+            {
+                fTnext = CLOCK_PERIOD * tpn;
+                break;
+            }
+        }
+    }
+    else
+    {
+        fR3BEventHeader->SetTprev(INVALID_EVENT);
+        fR3BEventHeader->SetTnext(INVALID_EVENT);
+        return;
+    }
+
+    if (dMScounter ==
+        0) // This if statement checks if there is no Hit correspondent to the MSHit even after the first if statement
+    {
+        fTprev = ERROR_NO_MS;
+        fTnext = ERROR_NO_MS;
+    }
+    if (dMScounter > 1)
+    {
+        fTprev = ERROR_MULTI_MS;
+        fTnext = ERROR_MULTI_MS;
+    }
+    if (dMScounter == 1)
+    {
+        if (sampHits == 1)
+        {
+            fR3BEventHeader->SetTprev(fOne_hit);
+            fR3BEventHeader->SetTnext(fOne_hit);
+            return;
+        }
+        if (fTprev < 0 && fTnext > 0)
+        {
+            fTprev = FIRST_LAST_HIT;
+        }
+        if (fTprev > 0 && fTnext < 0)
+        {
+            fTnext = FIRST_LAST_HIT;
+        }
```

Comment:
I would replace the last if block with:
```
else if (dMScounter == 1)
{
  if (fTprev == INVALID_TPTN)
     fTprev = fOne_hit;
  if (fTnext == INVALID_TPTN)
     fTnext = fOne_hit;
}
fR3BEventHeader->SetTprev(fTprev);
fR3BEventHeader->SetTnext(fTnext);
```

That way, whenever there is no previous/next hit, the "NoRecord" value will be used. For the special case of one hit (no previous and no next hit), both of these will automatically be set to that value.